### ISSUE TITLE
opinions please on approach to faster flipping

### ISSUE THREAD

**Issue by mpadge** [ropenscilabs/flipper#7]
@njtierney @MilesMcBain @LucyMcGowan @beemyfriend @kbenoit and anyone else: As Nick has previously noted, using `tools::CRAN_package_db()` can be frustratingly slow. Doing that plus processing all the texts through `quanteda` translates to the now pretty nifty `textsearch()` function taking up to 5 seconds. Issues:

1. That's way too long
2. The same procedures are repeated anew for each new `textsearch()`, which is obviously stupid

# solution

What I've tentatively done is port the `quanteda` processing of the `CRAN_package_db` texts to `.onLoad`. This is an admittedly unorthodox approach not without potential side effects. Please all: Have a play, see what you think
```
> library(flipper)
> textsearch("trees are rather nice") # or whatever
```

The good news? Runs in the blink of an eye (> 10 times faster) (And I know, it fails if you only enter one word - been concentrating too much on phrases me has. I'll fix that soon.)

Potential side effects:
1. The package will not load without an internet connection, which is stupid and dangerous.
2. Any other packages which depend on it will also not load without a ... (see above)
3. When there is an internet connection, loading can take quite some seconds, which is unexpected and unusual for a standard `library` call. Once again, any packages which depend on it will also be slowed down by the same amount just by loading -- not even attaching! - the namespace. Could be very frustrating for other package developers?

**Comment by nealrichardson** [comment]
Is there a reason not to defer the downloading and text processing to the first time a query is made? You're caching the results in a tempfile in the `.onLoad` version, so you could check for the existence of that file and only download if it doesn't exist. 

I would agree that requiring a network connection to load a package is unwise.

**Comment by mdsumner** [comment]
I use a similar approach  to put a large data frame into memory  on load,  it solves so many other problems I haven't looked back.  It could be done by sqlite query instead,and in your case I think an improvement would be a cache to fall back to when there is no Internet.  Buy overall I think it's  a good approach

**Comment by kbenoit** [comment]
A hybrid approach is to distribute the dataset with the package, but offer the user the option to refresh it, or simply to refresh the package on CRAN every few months. As a user I'd rather have speed than being 100% up to date.  (most packages are not going to change, just get updated.)

**Comment by MilesMcBain** [comment]
Check out the [memoise](https://cran.r-project.org/web/packages/memoise/index.html) package for caching function results. It works really well. You'll only pay the cost of downloading the CRAN db once per session. You could also memoise the search function.

**Comment by mpadge** [comment]
yeah, thanks @MilesMcBain - that is indeed going to be the solution. Ultimately via `memoise::cache_s3()`, because the plan is to cache what users actually do with it as well, in order to constantly update network weights
